src/execution_plan/ops/op_delete.c

```diff
@@ -189,50 +186,76 @@ static Record DeleteConsume(OpBase *opBase) {
 
 			SIValue_Free(value);
 		} else if(!(type & T_NULL)) {
-			/* Expression evaluated to a non-graph entity type
-			 * clear pending deletions and raise an exception. */
-			array_clear(op->deleted_nodes);
-			array_clear(op->deleted_edges);
-			// If evaluating the expression allocated any memory, free it.
+			// if evaluating the expression allocated any memory, free it.
 			SIValue_Free(value);
 			// free the Record this operation acted on
 			OpBase_DeleteRecord(r);
 			ErrorCtx_RaiseRuntimeException("Delete type mismatch, expecting either Node or Relationship.");
 			break;
 		}
 	}
-
-	return r;
 }
 
-static OpBase *DeleteClone(const ExecutionPlan *plan, const OpBase *opBase) {
-	ASSERT(opBase->type == OPType_DELETE);
-	OpDelete *op = (OpDelete *)opBase;
-	AR_ExpNode **exps;
-	array_clone_with_cb(exps, op->exps, AR_EXP_Clone);
-	return NewDeleteOp(plan, exps);
+static Record _handoff(OpDelete *op) {
+	return array_pop(op->records);
 }
 
-static OpResult DeleteReset(OpBase *opBase) {
+static Record DeleteConsume(OpBase *opBase) {
 	OpDelete *op = (OpDelete *)opBase;
+	Record r;
 
-	_DeleteEntities(op);
+	// return mode, all data was consumed
+	if(op->records) return _handoff(op);
 
-	if(op->deleted_nodes) {
-		array_clear(op->deleted_nodes);
-	}
+	// consume mode
+	op->records = array_new(Record, 32);
+	// initialize the records array with NULL
+	// which will terminate execution upon depletion
+	array_append(op->records, NULL);
 
-	if(op->deleted_edges) {
-		array_clear(op->deleted_edges);
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	ASSERT(op->op.childCount > 0);
+	// pull data until child is depleted
+	OpBase *child = op->op.children[0];
+	while((r = OpBase_Consume(child))) {
+		// persist scalars from previous ops before storing the record
+		// as those ops will be freed before the records are handed off
+		Record_PersistScalars(r);
+
+		// save record for later use
+		array_append(op->records, r);
+
+		// delete entities
+		_CollectDeletedEntities(r, opBase);
 	}
 
-	return OP_OK;
+	// done reading, we're not going to call consume any longer
+	// there might be operations e.g. index scan that need to free
+	// index R/W lock, as such free all execution plan operation up the chain.
```

Comment:
this comment is misleading: "as such free all " we're not freeing we're resetting